### ISSUE TITLE
RFC: Promotion Strategy

### ISSUE THREAD

**Issue by OmnipotentEntity** [leela-zero/leela-zero#1054]
There's been some discussion about how and when networks are promoted, and whether or not all networks ought to be promoted like in AZ, or if we should promote any network above 50%, or all sorts of other ideas.

We currently promote a little bit strange.  We want 95% confidence that the new network is better, so that's 55% at 400 games.  But we don't promote based on that directly.  We promote based on SPRT pass, which is 95% confidence that we will reach the 95% confidence, which is actually less than 95% total confidence.

Why not just play training games until the lower bound goes above 50% or the upper bound goes below 50%?  This might result in some networks going longer than 400 games, but that way there is less of a barrier to accepting better networks just because they can't quite hit 55% because their theoretical winrate might be only 51%.

However, in order to reach 95% confidence on a winrate of 51% (or worse 49%!) you need about 9000 games.  So... that's probably too many.

Another strategy might be similar to the above, but have a limit on number of games (400 is fine, but if we move to 1000 we can go down to 53%), and if the network fails to fail after that many games then the network copromotes and both networks generate games.  Then there's a question about which network should be the network to be tested against to become part of the promoted pool, and when networks should be retired.

Perhaps the testing flow might have each network represent the pool at random, and have the entire pool retired if a new network fully promotes rather than copromotes.  Or perhaps more interesting, on a full promote retire only the lowest winrate half, this would make full promotes actually copromotes, but function as a method of controlling the size of the pool when the differences in strength between the lower winrate members and the higher ones becomes high enough.

**Comment by gcp** [comment]
>We promote based on SPRT pass

That or 55% at the end. The SPRT pass is an early termination. In our current setup it seems mostly useful for bad networks, and in general the benefit is low.

>which is 95% confidence that we will reach the 95% confidence

No, it's about being more likely that it's 35 Elo better than that it's worse. (The distinction has been beaten to death in one of the issues here)

>Why not just play training games until the lower bound goes above 50% or the upper bound goes below 50%?

Because they use different parameters.

>the network fails to fail after that many games then the network copromotes and both networks generate games

And the benefit aside from making everything more complicated (and not actually being implemented anywhere) is...?

I can barely understand the rest of your post.

**Comment by OmnipotentEntity** [comment]
> Because they use different parameters.

I apologize, I meant match games.  I should have proof read more thoroughly.

> And the benefit aside from making everything more complicated (and not actually being implemented anywhere) is...?

The benefit would be not rejecting networks that are stronger just because they aren't strong enough for the difference to be easily and confidently measurable.  While still retaining networks that are nearly as strong just in case random chance promoted a weaker network.

It's not actually implemented anywhere; and empirical testing of this method is outside of the realm of my personal computational power (essentially, I'd have to replicate a scaled down version of LZ several times using different promotion strategies).  This is why I made an RFC.  To get some feedback on whether or not this might work or not, and get some people more experienced in CNN training to let me know whether or not this even makes sense.  I could perhaps run a few hundred thousand 5x5 training games with a small network in a few weeks or months and see how it worked out.

> I can barely understand the rest of your post.

I'll try to state it a bit more clearly in that case.

We have a pool of training networks, which are selected at random to produce a training game.  These training networks have associated ELO based on their match games when they were promoted, and also with newer networks before they are retired.

Periodically (how often?), new networks are trained from the pool.  Network generation strategy is just training the network with the highest ELO in the same manner as before.  (Note, that because networks can now change ELO based on newer match games the identity of the champion network can also change between training runs.)

The generated networks play match games against the pool.

If the new network has a strength of more than 35 ELO vs the pool's overall ELO (ie, if we treat the pool as a single opponent for the purpose of the match games against this network.)  Then add the network to the pool, sort by ELO and discard the lowest (proposed 50%, but other values may work, perhaps even 100%, or maybe spill until the mean of new_pool ELO >= mean of old_pool ELO + some value). (Promotion.)

If the new network has a strength of less than the pool's overall ELO, then the network is discarded.

If the new network has a strength between those, then add it to the pool. (Copromotion.)

Then repeat training.

(Note: Spilling the network on a 35 ELO difference is to 

Reasons why I think this might be beneficial:

1. More robust against potential "rock, paper, scissors" network behavior, because the new network doesn't have to face just one network, but all of them in the pool.
2. More variety in training games.  Potentially reducing the threat of overfitting.
3. Fails to throw away networks that are likely better, increasing the quality of training games on average in the slow growth tail.
4. More accurate ELO calculation, new networks aren't just being judged, they are also refining the relative placement of older networks and giving more confidence.

Reasons why I think this may be detrimental:

1. May slow progress in the beginning.  Because we are retaining some weaker networks on promotion, and because significantly stronger networks are easier to come by.
2. Complexity, as you mentioned this is more complicated, and the benefit isn't clear without testing. It's just conjecture at this point.

I hope this is more clear.  And I hope you can give me some feedback and insight or propose changes.

**Comment by klmsj** [comment]
I don't believe there is a "bad" network. As long as we generate the correct games, set the learning rate correctly, and so on, then the trained network must learn something, even if his winning percentage for the old network is low. This is easy to explain,  the winning percentage is low, indicating that the "skills" (or called  features,whatever ) of the network learned are negative.
There may be many factors that cause lead Go to win. When a network learns new "skills", it may be used frequently. If the "skills" are not effective, it will be suppressed in the next training cycle. Avoid using this "skills". Both positive and negative skills are very important. So I think the lz method is simpler and more effective.

**Comment by sethtroisi** [comment]
closing, no active discussion for ~1year with much discussion in other threads.
